docs/best_practice_guide.md

```diff
@@ -0,0 +1,127 @@
+# Best practice guide
+
+## What to use for
+
+You can use typo3-rector in various ways:
+- checking existing code if there are left out segments of the last upgrade
+- reformating e.g. typoscript files based on your editorconfig
+- evaluate upgrades and see what parts of your custom setup will be updated automatically
+- partial execution of full core upgrades
+
+
+
+
+## Guide for a good upgrade
+
+
+### Starting
+Starting with an upgrade should start with installing typo3-rector and checking for the rector rules/sets of your current version, not the one you're upgrading to.
+Often there are things that were missed out in previous upgrades while rector adds rulesets for those.
+
+One thing you should definitely need for upgrades with two or more version steps is the class-alias-maps.
+To provide the old classes and their new implementation typo3-rector add a ClassAliasMap.php for each version.
+Those can be added as single Map or for multiple versions at the same time in your composer.json. (mostly root composer, depending on your setup)
+
+```json
+"extra": {
+        "typo3/class-alias-loader": {
+            "class-alias-maps": [
+                "vendor/ssch/typo3-rector/Migrations/TYPO3/7.6/typo3/sysext/fluid/Migrations/Code/ClassAliasMap.php",
+                "vendor/ssch/typo3-rector/Migrations/TYPO3/8.7/typo3/sysext/fluid/Migrations/Code/ClassAliasMap.php",
+                "vendor/ssch/typo3-rector/Migrations/TYPO3/9.5/typo3/sysext/fluid/Migrations/Code/ClassAliasMap.php"
+            ]
+        },
+    },
+```
+
+So if you're on TYPO3 v8 you should start with applying the rulesets for v7 first and v8 afterwards to make sure you're current code base.
+
+Examples for often missed out upgrade steps:
+- ext:lang key changes OR the full ext:lang replacement
+- TCA renderType addition on type="single"
+- TCA fieldwizard OR overrideChildTCA rewrite
+
+
+### Ongoing upgrade
+
+After making sure your current code base is properly upgraded you go on with the actual upgrade process.
+This requires manual action like allowing the core versions in your composer.json and ext_emconf.php files depending on your individual setup.
+
+The overall package upgrade is then followed by some more typo3-rector actions.
+
+Depending on the amount of version steps you should add the ClassAliasMap as mentioned above for e.g. v8 if you're going from v8 to v10 directly.
+
+Once again you add you're wanted/needed rulesets that should be separated by version.
+It also comes in handy to divide between TCA and TYPO3 changes AND/OR you're packages
+
+In this config example we would apply all rules of both TCA and base TYPO3 to the package_numberone:
+
+```php
+    $containerConfigurator->import(Typo3SetList::TYPO3_95);
```

Comment:
TYPO3_95 already includes the TCA set for that version. So every TYPO3_* contains the corresponding TCA_* set

Reply:
ok. i'll rewrite that a little